extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +151,87 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticTimeoutsMap = new Map<URI, NodeJS.Timeout | undefined>();
+	private readonly _diagnosticSnapshotsMap = new Map<URI, readonly vscode.Diagnostic[]>();
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _getDiagnostics: (uri: URI) => readonly vscode.Diagnostic[]
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript') {
+				this._updateDiagnosticCodesAfterTimeout(e.document.uri, 10000);
+			}
+		}));
+		this._register(vscode.workspace.onDidOpenTextDocument(e => {
+			if (e.languageId === 'typescript') {
+				this._updateDiagnosticCodesAfterTimeout(e.uri, 10000);
+			}
+		}));
+		this._register(vscode.workspace.onDidCloseTextDocument(e => {
+			if (e.languageId === 'typescript') {
+				this._diagnosticTimeoutsMap.delete(e.uri);
+			}
+		}));
+		const activeUri = vscode.window.activeTextEditor?.document.uri;
+		this._updateDiagnosticCodesAfterTimeout(activeUri, 10000);
+		this._sendTelemetryEvent();
+	}
+
+	private _updateDiagnosticCodesAfterTimeout(uri: URI | undefined, timeoutInMs: number) {
+		if (!uri) {
+			return;
+		}
+		clearTimeout(this._diagnosticTimeoutsMap.get(uri));
+		const timeout = setTimeout(() => { this._updateDiagnosticCodes(uri); }, timeoutInMs);
+		this._diagnosticTimeoutsMap.set(uri, timeout);
+	}
+
+	private _updateDiagnosticCodes(uri: URI) {
```

Comment:
This seems to me to have the disadvantage of reporting only TS errors in opened documents. Would it be feasible to remove `uri` from here and maybe update here all TS diagnostics. I think when reading diagnostics from vscode.diagnostics it is possible to look at the owner and get only the ones for TS.

Reply:
Right, that makes sense. I changed the code so as to find the diagnostics for all the text documents in the workspace, initially in the constructor. Since I am calling the method getDiagnostics from the DiagnosticsManager, they are guaranteed to be TypeScript diagnostics.